users/talk.rs

```diff
@@ -0,0 +1,1240 @@
+                            break;
+                        }
+                    } else {
+                        // Handle connection closure
+                        handle_connection_close();
+                        break;
+                    }
+                }
+                Err(e) => {
+                    eprintln!("Error reading from stream: {}", e);
+                    break;
+                }
+            }
+        }
+    });
+
+    Ok(())
+}
+
+// Handles user input and displays it in the bottom section of the terminal.
+fn handle_user_input(
+    handle: &mut io::StdoutLock,
+    input: &str,
+    split_row: u16,
+    top_line: Arc<Mutex<u16>>,
+    bottom_line: Arc<Mutex<u16>>,
+) -> io::Result<()> {
+    let top_line = top_line.lock().unwrap();
+    let mut bottom_line = bottom_line.lock().unwrap();
+
+    // Reset bottom line position if it exceeds the available space.
+    if *bottom_line >= split_row - 2 {
+        *bottom_line = 0;
+    }
+
+    // Move cursor to bottom window
+    write!(handle, "\x1b[{};0H", split_row + *bottom_line + 1)?;
+    writeln!(handle, "{}", input)?;
+
+    // Move cursor to top window
+    write!(handle, "\x1b[{};H", *top_line)?;
+
+    handle.flush()?;
+
+    *bottom_line += 1;
+
+    Ok(())
+}
+
+// Draws the terminal interface with a split line separating the top and bottom sections.
+fn draw_terminal(split_row: u16, width: u16) -> io::Result<io::StdoutLock<'static>> {
+    let stdout = io::stdout();
+    let mut handle = stdout.lock();
+
+    // Clear terminal screen
+    print!("\x1B[2J\x1B[H");
+    io::stdout().flush().ok();
+
+    // Display the connection message at the top of the terminal.
+    write!(handle, "[Connection established]")?;
+    // Move the cursor to the split row and draw the split line.
+    write!(handle, "\x1b[{};0H", split_row)?;
+    // Draw the horizontal split line (─) across the width of the terminal.
+    writeln!(
+        handle,
+        "└{:─<width$}┘",
+        "",
+        width = (width as usize).checked_sub(2).unwrap_or(0)
+    )?;
+    // Move the cursor back to the top-left corner and then down by one line.
+    write!(handle, "\x1b[1;H")?;
+    write!(handle, "\x1B[1B")?;
+
+    handle.flush()?;
+
+    Ok(handle)
+}
+
+// Opens a TCP socket bound to the provided IPv4 address.
+// Used for talk users sending and receiving messages.
+fn open_sockt(my_machine_addr: Ipv4Addr) -> Result<(SocketAddr, TcpListener), io::Error> {
+    let listener = TcpListener::bind((my_machine_addr, 0))?;
+    let addr = listener.local_addr()?;
+
+    Ok((addr, listener))
+}
+
+// Opens a UDP socket bound to the provided IPv4 address.
+// Used for checking properly access to connection.
+fn open_ctl(my_machine_addr: Ipv4Addr) -> Result<(SocketAddr, UdpSocket), io::Error> {
+    let socket = UdpSocket::bind((my_machine_addr, 0))?;
+    socket.set_nonblocking(true)?;
+    let addr = socket.local_addr()?;
+
+    Ok((addr, socket))
+}
+
+fn handle_connection_close() {
```

Comment:
If you've changed to the raw mode or will after the previous comment - don't forget to return to normal after.